Battery Checker

calvinwt opened the pull request:
Simple program to check the voltage and input current of the battery, and then print it to shuffleboard. Also has the capability to run a timer when the robot first starts, changing a boolean when it ends.

calvinwt committed:
one

caseyjbrotherton commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -16,6 +16,8 @@
  */
 public final class Constants {
 
+  public static final int minvoltage = 10;
```
Feel like this should start higher.  at 10.5 mininum, or 11.

from what I am seeing on ChiefDelphi, 10.5V is seen as the point where battery performance drops quickly to zero.
And we don't want the batteries to be harmed by draining them. 

 ( And because we can't shut down the robot, we should assume that there will be some current drop while a team member walks onto the field to switch the battery. )

https://www.chiefdelphi.com/t/about-battery-life/126825
https://www.chiefdelphi.com/t/cba-5-and-battery-testing/408305

calvinwt replied:
Completed, changed it to 12 for good measure

caseyjbrotherton commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,78 @@
+  private Timer timer;
+
+  private void init() {
+    timer.reset();
```
timer = new Timer();

This code only counts the time since the robot code reset the timer.  Can we get since the robot was turned on?

calvinwt replied:
Completed, moved the reset to robotInit()

caseyjbrotherton commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,78 @@
+
+  public void periodic() {
+    sbVoltage.setDouble(getVoltage());
+    sbInCurrent.setDouble(getInputCurrent());
```
We would like to track high current usage.  Can we accumulate times of high current usage?
It would be ok to count "ticks"  approximately 20ms intervals...
But it would be even better to accumulate the real time in between ticks.

calvinwt replied:
Completed, made it delay the timer if the current is below a certain value (still need to find that value though)

caseyjbrotherton commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,78 @@
+    sbVoltage.setDouble(getVoltage());
+    sbInCurrent.setDouble(getInputCurrent());
+    sbTimer.setDouble(getTimer());
+    sbTimerChange.setBoolean(
```
We want to warn if any of the three conditions happen:
1.  High current usage for some amount of time
2. Any usage for a longer time
3. Dangerously low voltage. 

It would also be great to have a "yellow zone" of optional alert, and then have a more forceful, "danger zone" alert.
Maybe even disable the robot in the danger zone?

Maybe yellow zone is 1 or 2?  and danger zone is passing 3?  Not sure what is best.

caseyjbrotherton replied:
I should note, not sure that one can disable the robot programatically.  But I feel like we need to try and consider a way to make it impossible to ignore, if we can

Is there something more annoying that we can do in the danger zone?  Continuous beeping on the robot?

calvinwt replied:
Completed, yellow and red zone exist only in the actual code, the simulation only uses 1 marker for checking times

caseyjbrotherton commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,78 @@
+  }
+
+  public boolean checkTimer() {
+    if (timer.hasElapsed(300)) {
```
Let's define 300 as a constant.

calvinwt replied:
Completed

caseyjbrotherton commented on src/main/java/frc/robot/commands/BatteryCommand.java:
```diff
@@ -0,0 +1,34 @@
+
+  @Override
+  public void execute() {
+    subsystem.periodic();
```
subsystem.periodic s run once per run of the command scheduler already?

https://docs.wpilib.org/en/stable/docs/software/commandbased/subsystems.html

Don't think we need this command.

calvinwt replied:
Completed, removed it

caseyjbrotherton reviewed (dismissed):
Good start.  I noted some places that need some tweaks and more thought.

Don't think that we need the default command, but might need to create other commands.

There is another pattern besides what we used last season.  We will be talking about this further in our meetings.  Here is a pre-release document from WPIlib talking about the other pattern:
https://github.com/wpilibsuite/frc-docs/blob/bc87e1e906e608c6517399e5cb6fa45c11eabd42/source/docs/software/commandbased/organizing-command-based.rst

If we need another command, instant commands and instant command factories might be a really good pattern for the Battery Checker.

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,78 @@
+  }
+
+  public void periodic() {
+    sbVoltage.setDouble(getVoltage());
+    sbInCurrent.setDouble(getInputCurrent());
```
Shuffleboard boolean is good, but you might think about also using [DriverStation.reportError](https://first.wpi.edu/wpilib/allwpilib/docs/release/java/edu/wpi/first/wpilibj/DriverStation.html#reportError(java.lang.String,boolean))

calvinwt replied:
Completed

NogaRoze committed:
addressing comments

NogaRoze committed:
addressing more comments

calvinwt committed:
improved warning systems

calvinwt committed:
Merge branch 'BatteryTest' of https://github.com/AusTINCANsProgrammingTeam/2022RobotCode into BatteryTest

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,120 @@
+      DriverStation.reportError("Change the Battery!", true);
+    } else if (getVoltage() < Constants.minVoltageYellow) {
+      DriverStation.reportError("Change the Battery Soon!", true);
+    }
```
Should the "Change the Battery Soon!" cases be reportWarning instead of reportError? 

Also, I think printing the stack trace is unnecessary (boolean parameter of reportError/Warning) and probably will hide the actual message from the operator. 

Also also, once you hit an error case is the error going to reprint every 20 ms forever?

calvinwt replied:
TBD

calvinwt replied:
completed, moved to warning and stopped stack trace from printing

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,120 @@
+  }
+  public void checkCurrent() {
+    if(getInputCurrent() < Constants.maxBatteryCurrent) {
+      Timer.delay(0.02); //Was current timer, but I got a warning about it needing to be static so I changed it to this. Will this cause conflicts?
```
We should let the scheduler handle delays and not use this. You can be certain that every call of periodic() happens 20 ms after the last. 

Plus I don't think this delay helps anything. I think the intent was to count the times we read a high current (or start/stop currentTimer to be somewhat more accurate), and then if the battery has experience high current draw for too long since being changed display the error/warning.

calvinwt replied:
Done, switched to start() and stop() functions

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,120 @@
+
+
+  public double getTimer() {
+    return Timer.getFPGATimestamp();
```
This gets the system clock time, and isn't specific to the Timer's you've created. What you want to use is 
timer.get() and currentTimer.get() to get the times from those specific timers.

calvinwt replied:
completed

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  public BatterySubsystem() {
+    init();
```
Is there a reason we need a separate private method for the constructor?

calvinwt replied:
fixed to just initialize when called

azvanderpas reviewed (changes requested):
I think before anything else we should try to setup simulation and/or unit tests for this. Seems like a good project to learn those skills on.

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,120 @@
+  }
+
+  public boolean checkTimer() {
+    if (currentTimer.hasElapsed(Constants.timeInSecondsHighCurrentRed)) {
+      DriverStation.reportError("Change the Battery!", true);
+      return true;
+    } else if (currentTimer.hasElapsed(Constants.timeInSecondsGeneralRed)) {
+      DriverStation.reportError("Change the Battery!", true);
+      return true;
+    } else if (timer.hasElapsed(Constants.timeInSecondsHighCurrentYellow)) {
+      DriverStation.reportError("Change the Battery Soon!", true);
+      return false;
+    } else if (timer.hasElapsed(Constants.timeInSecondsGeneralYellow)) {
+      DriverStation.reportError("Change the Battery Soon!", true);
+      return false;
+    }
+    return false;
```
Each time we redeploy code these timers will be reset. This may not true of Timer.getFPGATimestamp(), but that is somewhat unclear. Either way tracking time elapsed across redeploys presents a problem.

calvinwt replied:
TBD

calvinwt replied:
done, timer values are now saved to the robot with Preferences

calvinwt committed:
fixed some init errors and cleaned up code

calvinwt committed:
reformat text

calvinwt committed:
simulate voltage

calvinwt committed:
make color swatch change color

NogaRoze committed:
Co-authored-by: calvinwt <calvinwt@users.noreply.github.com>

azvanderpas committed:
Add JUnit dependancy in order to use

assertEquals in BatterySimulation.
Assert keyword is disabled by default, and gradle configuration is too
opaque for me to figure out how to enable it

calvinwt committed:
add multiple simulation functions as well as print statements to debug

NogaRoze committed:
Co-authored-by: calvinwt <calvinwt@users.noreply.github.com>

NogaRoze committed:
Addressed high current timer and warning system

NogaRoze committed:
Made robot turn itself on

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,221 @@
+    return RobotController.getInputCurrent();
+  }
+
+  public void checkCurrent() {
```
This method does nothing...

calvinwt replied:
Fixed, current code for some reason didn't push properly

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySimulation.java:
```diff
@@ -0,0 +1,65 @@
+      DriverStationSim.setEnabled(true);
+    }
+
+    //Forcefully sets the voltage when enabling the robot.
+    if (DriverStationSim.getAutonomous() == true || DriverStationSim.getEnabled() == true) {
+      batterySubsystem.startHCTimer();
+      //RoboRioSim.setVInVoltage(12.1);
+    } else {
+      //RoboRioSim.setVInVoltage(14.0);
+      batterySubsystem.stopHCTimer();
+    }
```
The timers need to be controlled within the BatterySubsystem itself. The only reason to start or stop the timers in simulation is to inject some sort of error case you want to catch. 

This code won't run on the robot

calvinwt replied:
Fixed, changed it to change sim current

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,221 @@
+    if (Robot.isSimulation()) {
+      return (int) simVolt;
+    } else {
+      return RobotController.getBatteryVoltage();
```
I don't think you need a different return value for simulation. Since we are using the RoboRioSim to set the voltage it should get returned properly here.

calvinwt replied:
done, removed the variable

calvinwt replied:
changed it to return with a RoboRioSim.getVInVoltage() method, should I just use the getBatteryVoltage() one instead?

azvanderpas replied:
Yeah, I think just using RobotController.getBatteryVoltage() works in simulation and in real

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,221 @@
+        btTab.add("High Current Timer", 0).withSize(2, 2).withPosition(0, 4).getEntry();
+    sbTimerChange = btTab.add("Change Timer", 0).withSize(2, 2).withPosition(4, 0).getEntry();
+    timer.start();
+    currentTimer.start();
```
The high current time probably shouldn't start automatically

calvinwt replied:
fixed

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,221 @@
+      // playAudio(Constants.audiofilepath);
+      return false;
+    } else if (timer.hasElapsed(Constants.timeInSecondsGeneralYellow)) {
+      DriverStation.reportWarning("Change the Battery!", false);
```
Red and yellow messages are no longer distinguishable

calvinwt replied:
fixed, messages are now distinguishable

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySimulation.java:
```diff
@@ -0,0 +1,65 @@
+  }
+
+  @Override
+  public void simulationPeriodic() {
```
The different tests should be sequenced in a better way. Since the whole method runs each 20 ms loop there are probably some conflicting conditions happening. Using JUnit testing framework may make some of this easier, but using getFPGATimestamp() to only run tests over a specific interval of runtime would also work. Something like this:
```
double now = Timer.getFpgaTimestamp();
if (now >= 0 && now < 1.0) {
    // Do voltage check test
} else if (now >=1.0 and now < 30.0) {
    // Do timer test
}
// else if (etc....)
```

calvinwt replied:
fixed, timing is now better

calvinwt committed:
streamlined testing process and fixed some redundant code regarding simulation voltage

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySimulation.java:
```diff
@@ -27,39 +20,34 @@ public BatterySimulation(BatterySubsystem batterySubsystem) {
+        System.out.println("Got to assert general timer");
+        assertEquals(true, batterySubsystem.checkTimer());
+      }
+    } else if (now >= 25.0 && now < 50.0) {
```
Tie time durations to your constants. Your change to timeInSecondsGeneralRed breaks this test now.

calvinwt replied:
Fixed, tied to constants and reorganized for better flow

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySimulation.java:
```diff
@@ -27,39 +20,34 @@ public BatterySimulation(BatterySubsystem batterySubsystem) {
 
-    //Forcefully sets the voltage when enabling the robot.
+    // Forcefully sets current when robot is enabled to simulate use
     if (DriverStationSim.getAutonomous() == true || DriverStationSim.getEnabled() == true) {
```
What is the autonomous check for?

calvinwt replied:
autonomous check removed, it was unnecessary

azvanderpas commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -17,14 +17,16 @@
+  public static final double timeInSecondsHighCurrentRed = 20.0;
+  public static final double simCurrentLow = 10.0;
+  public static final double simCurrentHigh = 13.0;
   public static final double maxBatteryCurrent = 12.3; // TODO: Put a real value here
```
I'd suggest a better name. Max current implies it isn't safe to operate above this current. 
How about highCurrentBatteryThreshold?

calvinwt replied:
done, renamed to highCurrentBatteryThreshold

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySimulation.java:
```diff
@@ -27,39 +20,34 @@ public BatterySimulation(BatterySubsystem batterySubsystem) {
     } else {
-      //RoboRioSim.setVInVoltage(14.0);
-      batterySubsystem.stopHCTimer();
+      RoboRioSim.setVInCurrent(Constants.simCurrentLow);
```
The constants here seem a little unnecessary since you already have a constant for what is high current. You might do something like RoboRioSim.setVInCurrent(Constants.maxBatteryCurrent+1) for high current and -1 for low.

calvinwt replied:
done

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,223 @@
+  }
+
+  public boolean checkRedVoltage() {
+    if (getVoltage() > Constants.minVoltageRed) {
+      return true;
+    } else {
+      return false;
+    }
```
```suggestion
    return getVoltage() > Constants.minVoltageRed;

```

calvinwt replied:
done

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,223 @@
+    btTab = Shuffleboard.getTab("Battery");
+    // Change 16 to desired value
+    Shuffleboard.getTab("Battery")
+        .addBoolean("Battery Voltage", () -> getVoltage() > Constants.minVoltageRed)
```
```suggestion
        .addBoolean("Battery Voltage", () -> checkRedVoltage())
```

calvinwt replied:
done

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,223 @@
+    return currentTimer.get();
+  }
+
+  public void startHCTimer() {
+    currentTimer.start();
+  }
+
+  public void stopHCTimer() {
+    currentTimer.stop();
+  }
```
I don't think we want to expose the timer controls as public. Either make these private or better just use currentTimer directly within this class and delete these.

calvinwt replied:
done

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,223 @@
+    return false;
+  }
+
+  /*
+  public void checkMinute() {
+    double check = (getTimer() % 60.0);
+    if (check > 0 & check < 0.025 & getTimer() > 1.0) {
+      minutesPassed += 1;
+    } else {
+    }
+  }
+  */
+
+  public int getMinute() {
+    return (int) Math.floor(timer.get() / 60.0);
+  }
```
Clean up unused methods and commented out code

calvinwt replied:
done

NogaRoze committed:
Replaced RIO with Power Distribution

calvinwt committed:
made system more efficient and improved information relay. Shuffleboard now properly shows current and voltage, as well as accurate timer values

calvinwt committed:
Reorganized shuffleboard and added more color indicators for both timers

calvinwt committed:
Added comments for better readability

calvinwt committed:
Addressed PR comments and made timer values save to roborio

calvinwt committed:
built to pass clean check on github

caseyjbrotherton committed:
Update main.yml

Debug workflow file

calvinwt committed:
changed gitignore to ignore networktables.ini changes (was causing errors with github checks)

calvinwt committed:
Merge branch 'BatteryTest' of https://github.com/AusTINCANsProgrammingTeam/2022RobotCode into BatteryTest

azvanderpas commented on src/main/java/frc/robot/RobotContainer.java:
```diff
@@ -153,6 +167,10 @@ private void initCommands() {
       driveBaseTeleopCommand = new DriveBaseTeleopCommand(driveBaseSubsystem);
       driveBaseSubsystem.setDefaultCommand(driveBaseTeleopCommand);
     }
+    // if (batterySubsystem != null) {
+    //  batteryCommand = new BatteryCommand(batterySubsystem);
+    //  batterySubsystem.setDefaultCommand(batteryCommand);
+    // }
```
Doesn't look like BatteryCommand is used and everything is done in periodic. Remove the class and references to it.

calvinwt replied:
done, command has been removed

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,199 @@
+    return currentTimer.get() + storedHighCurrentTime;
+  }
+
+  private void startHCTimer() {
+    currentTimer.start();
+  }
+
+  private void stopHCTimer() {
+    currentTimer.stop();
+  }
```
Methods usually only need to be created to add a capability/access to the object from an outside source (public methods), or to reuse code that is commonly used and relatively long. Since this case is neither, I would rather just use currentTimer.start and stop directly where we needed it instead of having methods for them.

calvinwt replied:
done, methods have been removed

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,199 @@
+    return getVoltage() > Constants.minVoltageRed;
+  }
+
+  public void resetTimers() {
```
Good work figuring the Preferences out!
One addition I would make is to save the last 'low current' voltage as a Preference. 
Then if you read a higher voltage on startup, reset the timers because the battery has likely been changed.

calvinwt replied:
done, but still needs testing

calvinwt replied:
did some testing, seems like no matter what, if you reboot the roborio it clears the values of preferences. Either way, performance should be as intended

azvanderpas commented on src/main/java/frc/robot/subsystems/BatterySubsystem.java:
```diff
@@ -0,0 +1,199 @@
+    currentTimer.reset();
+  }
+
+  // TODO: Find a way to make these not reset every redeploy. Possibly make them save their values
+  // to a file, then
+  // retrieve that file when the robot starts.
```
Looks like you've solved this already. Delete comment

calvinwt replied:
done

azvanderpas reviewed (dismissed):
Will approve with the clean up tasks. Battery changed checking can go in a new PR

Commit:
completed PR edits, removed battery command, added competition constant to toggle battery subsystem with

Commit:
built to pass spotless check

Commit:
did some testing on timer resets (stuff regarding voltage tests has been commented out, may be implemented later)

calvinwt committed:
built to pass spotless check, added pathplanner vendor library for later

calvinwt committed:
Cleaned up comments in BatterySubsystem.java, added clarifying comments to timer alerts

calvinwt committed:
built to pass spotless check
